proposals/0148-stake-program-move-instructions.md

```diff
@@ -0,0 +1,201 @@
+becomes a second active stake account delegated to V1 with `amount` stake.
+* `Deactivate` T and wait an epoch.
+* `DelegateStake` T to V2 and wait an epoch. T becomes an active stake account
+delegated to V2 with `amount` stake.
+* `MoveStake` the `amount` stake from T to U2. T returns to its initial inactive
+state. Stake has moved from U1 to U2 with no outside lamports required and no
+new undelegated lamports becoming trapped in delegated stake accounts.
+
+The motivation for `MoveLamports` is to enable housekeeping tasks such as
+reclaiming lmaports from `Merge` destinations.
+
+## Alternatives Considered
+
+* There is a longstanding proposal we call Multistake, which would allow a
+stake account to have two delegation amounts to the same validator, the idea
+being that one serves as an onramp (and possibly offramp) for active stake on
+the account. This could support other flows to accomplish the same objective,
+such as allowing `DelegateStake` to accept an active stake account to begin
+activating an account's excess (non-rent non-stake) lamports, or allowing a
+`Split` source to `Deactivate` enough stake to cover rent-exemption for the new
+account. However, Multistake is a much larger design/engineering project, and
+we have to solve this sooner than it would be ready.
+* We discussed various proposals for allowing `Merge` to leave behind the
+source account or `Split` to split into any mergeable destination. However this
+confuses the presently clear distinction between these two operations and
+entails additional implementation risk as they are already rather complex. A
+new instruction that does one specific thing seems highly preferable.
+* The original version of this SIMD proposed a `Move` that did not take an
+`amount`, but this would require changes to `Split` to enable the first leg of
+the proposed flow.
+* Back out the changes introduced by requiring rent-exempt `Split` destinations.
+This is undesirable because that restriction was added for very good reason: an
+effectively unbounded amount of stake could be instantly deactivated through
+repeated splitting.
+
+## New Terminology
+
+`MoveStake` and `MoveLamports`, two new stake program instructions.
+
+## Detailed Design
+
+For clarity of terminology inside this specification:
+
+* An "active" stake is in a `Stake` state with nonzero delegation, 100% of which
+is effective stake. There is no activating or deactivating stake.
+* An "inactive" stake is in an `Initialized` or `Stake` state. There is no
+effective, activating, or deactivating stake.
+
+### `MoveStake`
+
+`MoveStake` requires 5 accounts:
+
+0. Source stake account: Writable, owned by the stake program
+1. Destination stake account: Writable, owned by the stake program
+2. Clock: Read-only, with the fixed address
+`SysvarC1ock11111111111111111111111111111111`
+3. Stake history: Read-only, with the fixed address
+`SysvarStakeHistory1111111111111111111111111`
```

Comment:
Why do we need to pass these in as accounts?

Reply:
hm we dont need to. i just thought it would be best practice since most instructions do this, even if they arent used. should i remove them?

incidentally when the bpf stake program comes, i was planning on making none of the instructions require sysvars anymore and just skipping them if provided

Reply:
I think remove them. We do have some stake-program instructions (like DeactivateDelinquent) already forgoing sysvar accounts, so no consistency reason to keep them. Might as well make these new instructions take up less transaction space from the get-go.